UACloudLibraryServer/Controllers/InfoModelController.cs

```diff
@@ -187,16 +187,18 @@ public async Task<IActionResult> DownloadNamespaceAsync(
                 return new ObjectResult("Could not parse identifier") { StatusCode = (int)HttpStatusCode.BadRequest };
             }
 
-            if (nodesetXMLOnly)
+            UANameSpace uaNamespace = await _database.RetrieveAllMetadataAsync(nodeSetID, User.Identity.Name).ConfigureAwait(false);
+            if (uaNamespace == null)
             {
-                await _database.IncrementDownloadCountAsync(nodeSetID).ConfigureAwait(false);
-                return new ObjectResult(nodesetXml) { StatusCode = (int)HttpStatusCode.OK };
+                return new ObjectResult("Failed to find nodeset metadata") { StatusCode = (int)HttpStatusCode.NotFound };
             }
 
-            UANameSpace uaNamespace = await _database.RetrieveAllMetadataAsync(nodeSetID).ConfigureAwait(false);
-            if (uaNamespace == null)
+            // if we got here, the user is allowed to download the nodeset
+
+            if (nodesetXMLOnly)
             {
-                return new ObjectResult("Failed to find nodeset metadata") { StatusCode = (int)HttpStatusCode.NotFound };
+                await _database.IncrementDownloadCountAsync(nodeSetID).ConfigureAwait(false);
+                return new ObjectResult(nodesetXml) { StatusCode = (int)HttpStatusCode.OK };
             }
 
             uaNamespace.Nodeset.NodesetXml = nodesetXml.Blob;
```

Comment:
this line will crash if metadataOnly is true
It should be moved to the next !metadatOnly if block

Reply:
Fixed.

Reply:
Fixed.